qualtran/bloqs/factoring/mod_arithmetic.py

```diff
@@ -0,0 +1,224 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Dict, Optional, Set, Union
+
+import sympy
+from attrs import frozen
+
+from qualtran import Bloq, Register, Signature
+from qualtran.bloqs.basic_gates import Toffoli
+from qualtran._infra.composite_bloq import BloqBuilder, SoquetT
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class ModAdd(Bloq):
+    r"""An n-bit modular addition gate.
+
+    Implements $U|x\rangle|y\rangle|p\rangle \rightarrow |x\rangle|y + x mod p\rangle|p\rangle$ using $4n T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        y: A bitsize-sized input/output register (register y above).
+        p: A bitsize-sized input register (register p above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('x', bitsize=self.bitsize),
+                Register('y', bitsize=self.bitsize),
+                Register('p', bitsize=self.bitsize),
+            ]
+        )
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(4 * self.bitsize, Toffoli())}
+
+    def build_composite_bloq(
```

Comment:
Should probably remove the empty build_composite_bloq methods since the base bloq class will raise an error if it's called without an implementation.